### ISSUE TITLE
Verify fingerprint in Webcrypto demo

### ISSUE THREAD

**Issue by leplatrem** [Kinto/kinto-signer#79]
@Natim suggested in #76 to use the `public_key` attribute instead of parsing the X.509 PEM.

This would drastically simplify the signature verification since we could get rid of the `x509ecdsa.js` file :)

**Comment by Natim** [comment]
According to @jvehent code here: https://github.com/mozilla-services/autograph/blob/master/tools/autograph-client-python/client.py#L59-L61

The key is still in a PEM format, however I am not sure of how we should load it to get X and Y values.

**Comment by Natim** [comment]
Using http://bit.ly/1VUxdMa shows a part of http://bit.ly/1ZF3UeK

We should be able to plug it here: https://github.com/Kinto/kinto-signer/blob/master/demo/x509ecdsa.js#L244

**Comment by jvehent** [comment]
I'm not 100% sure of what the end goal is, so this may be off the mark but here it is.

Content-Signature has two mode of operations:
1. the public key is passed to the client in an `Encryption-Key` http response header, and the key id is mapped in the `Content-Signature` response header. This mode is not implemented in Firefox.
2. the public key is stored somewhere and its location is passed through the `x5u=` value of the `Content-Signature` response header. This mode is implemented in Firefox.

If you want to use mode #1, then I would suggest, in the kinto backend, to detect when the x5u value is **not** set. If that's the case, take the public key, serve it through an `Encryption-Key` response header to the client, and then process it.

Currently, autograph returns the public key in DER-encoded PKIX format. The Content-Signature spec does not indicate which format the `Encryption-Key` should be served as, but it's possible the intention was to not use DER and just concatenate the X and Y values like we do for signatures. @franziskuskiefer would know more.

**Comment by franziskuskiefer** [comment]
We don't support case 1 and I'm not sure if we should. What's the point of having the key sent along with the signature? We implement the key-id version though (til I remove it next week).
The format of Encryption-Key is not defined (like many things) but it should probably NOT be DER :)
So what exactly is the use case here? I'd in general strongly argue against using Encryption-Key.

**Comment by leplatrem** [comment]
The use-case is to have a simple demo that verifies an Autograph signature (ECDSA P384) using WebCrypto.

We have some [client code that will ship in Gecko](https://bugzilla.mozilla.org/show_bug.cgi?id=1263602), but it relies on NSS and we would like to generalize it with WebCrypto.

Tim and Mark confirmed us that it was not possible to import a ECDSA key from its SPKI with `crypto.subtle.importKey`. We had to go for JWK, which requires `x` and `y` to be provided as base64url.

In [the current demo](https://kinto.github.io/kinto-signer/), we parse the cert chain (.pem from x5u) to extract the public key from its numerous fields. It is rather fragile and [a lot of code](https://github.com/Kinto/kinto-signer/blob/6211791/demo/x509ecdsa.js) is dedicated to X.509 and ASN1 parsing.

Switching from parsing the x5u to parsing the `public_key` would be less fragile, but would still require the same amount of boilerplate code for parsing if it comes as DER. Same applies to the `Encryption-key header.

I realize that we could also parse the public key in the kinto backend and augment the response we get from autograph with the two fields `x` and `y`, just ready to be consumed by clients using WebCrypto.

```python
key = response['public_key']
x, y = parseDER(key)
response['public_key'] = {'der': key, 'x': x, 'y': y}
return response
```
We are not familiar with every aspect of this task, but find it nice to have a simple illustration of how to validate a signature with JS in the readme of this repo :) https://github.com/Kinto/kinto-signer#validating-the-signature

Thanks a lot for your feedback and support :)

**Comment by jvehent** [comment]
It would be easy to return the public key in the autograph response in a format that's easy to parse without involved DER. Probably way easier than you parsing it in kinto backend, and also nicer because I really think kinto backend should not attempt to do any crypto at all, including parsing stuff.

That said, the security model isn't great. If you just use the public key, you don't check the chain of trust that guarantees the signature came from a certificate signed by a trusted authority. It makes the content signature pretty much useless because any CDN that relays your data can just replace the encryption key and content signature with its own. I'm very skeptical there is any security benefit in using this security model. Worse, there is a risk people would misuse it thinking it makes things safer when it doesn't. But maybe I'm missing something...

**Comment by franziskuskiefer** [comment]
I agree with Julien. There is no security benefit in doing this and doing an example this way only makes people use it.

But maybe we can get something else working. With key-id you can specify an ID to look up a public key with. We currently do this with a hard coded pref. But maybe adding this to the verification service in Firefox with something customisable (pass a pref to the service where it can then look for the key) could work. However I'm not sure if that's really easier than using the PKI. But it would allow people to play with it without getting a PKI set up and without having to do all the parsing.
I'm not sure if I understand the idea of using only WebCrypto here. The problem is that you'll need a trust anchor or a "pinned" key to establish trust. Something that's a bit difficult with WebCrypto only.

**Comment by jvehent** [comment]
Unfortunately, Content Signature is not usable safely outside of Mozilla right now. The only way to make it secure is to use the PKI we hardcoded in Firefox, or hardcode _something else_ in Firefox. Maybe an addon could do it, but that's hardly something we can recommend.

I think we should be careful about what we put out there in terms of technological demos. The security world can be very critical of imperfect solutions, particularly when they come from an organization like Mozilla that's expected to be at the top of the crypto world. We should make sure that whatever we publish to showcase Content Signature provides high security. That said, it shouldn't stop us from experimenting with stuff, just make sure it doesn't claim to provide any security.

In the future, I think we should allow users to sign with certificates that have roots in the truststore. In the current model, it's just a matter of using a Code Signing certificate with the right SAN records. We could still pin roots with HPKP. It's a path for us to standardize Content Signature and make it publicly available, but will take time. And it also doesn't solve your attempt at not parsing X.509 certificates :disappointed:

**Comment by Natim** [comment]
I'd like to provide a simple and secure solution for user that will need to use kinto with the collection signature feature. There are no reason if at Mozilla we need it other won't.

Having the collection signed with a public key pinned on the client side seems like a good first step while it makes it harder to update/rotate the key later.

Reading what you explained here, it probably means validating the chain of trust using WebCrypto and telling people how to correctly setup the chain of trust and pin their root certificate on client side.

Crypto is hard, we make mistakes but the same goes with OpenSource software and we need to educate and iterate on the solutions we provide (because they won't be perfect at first) to make it easier for people to implement secure solution for their use cases.

**Comment by leplatrem** [comment]
Thanks everyone for your feedback!

The worst is: I didn't realize that what I was proposing was completely flawed until you pointed it out :)

Well, not completely, it does not guarantee authenticity of data but it still guarantees integrity between client and server data.

If I understand correctly, since we can't prove who authored the public key, the easiest way is to hard-code it in the client. This is a limitation, but along with your explanations it can be understandable.

As for the «exemplariness», I don't think we had too much pretention yet. The day before yesterday I didn't know what a PKI was ;)

**Comment by jvehent** [comment]
@Natim: Your understanding is correct. However, we don't have an interface in Firefox today that will allow people to add a Root CA to the Content-Signature verifier, so we can't yet tell other people to use it. @mozmark may ideas in that area.

@leplatrem: Hardcoding in the client is a hard problem because it has to be done before the client first connects to the site. In Firefox, we control the platform, so we can do it ourselves. For others, than means relying on certificate authorities that are hardcoded and allowed to issue signing certificates.

Side note, for crypto knowledge, @marumari's slides (250 of them!) are good. https://marumari.github.io/crypto-presentation/

**Comment by jvehent** [comment]
> Well, not completely, it does not guarantee authenticity of data but it still guarantees integrity between client and server data.

It doesn't. Without authentication, you can't have integrity. It's a core principle of cryptography.
Anyone with the ability to modify traffic between the client and the server can replace the data, signature and public key with their own.

In its current form, the demo provides less security than TLS itself, because TLS, at least, verifies the domain owners have not been hijacked (through chain of trust verification). Since TLS already protects the communication between the client and the server, there is no security benefit to add a layer of content signature that doesn't provide authentication.

Sorry for being a downer. I do think this is a very cool demo. Once we have the ability to open the Root CAs to users, it will provide a lot of security. It allows anyone to serve their data through multiple layers of relays (CDNs) without worrying about integrity loss, and doing the verification in Javascript with WebCrypto is cool and visual. We just don't have the pieces in place to allow you to use it securely right now :disappointed:

**Comment by Natim** [comment]
So if I understand correctly pinning in Javascript won't be possible except if we could have Javascript signed file and then we would need the PKI to validate the Javascript signature.

Thanks a lot for explaining all that :)

**Comment by franziskuskiefer** [comment]
So, back from PTO and more than a phone to type on...

I think you can still do a useful example in JS only to demo this. The only problem is the authenticity, right? How about adding an input field that allows to paste a certificate fingerprint for either the signer certificate (if encryption-key is given) or the root of the trust chain (if x5u is given) and then verify (in addition to the signature and the cert chain) that the fingerprint matches. That is essentially what Firefox is doing as well and shouldn't be too much additional work.
It's not as nice as before as it requires interaction in the verification process. But as long as there is no root store that's the best way you can make this example useful imho.
